### PR TITLE
Convert sub to add and div to mul when doing the reduction across the per-thread reduction array.

### LINKED ISSUE #4555: incorrect result with parfor and -=
Pardon my sloppy numpy. All we are doing is computing `0 - 1 - 2 - 3 - ... - 9`.

```py
import numpy as np
from numba import njit, prange

@njit(parallel=True)
def two_d_array_reduction_prod_parallel(n):
    shp = (10, 10)
    result1 = np.zeros(shp, np.int_)
    tmp_shp = (n, 10, 10)
    tmp = np.ones(tmp_shp, np.int_)

    for i in range(n):
        tmp[i] = i

    for i in prange(n):
        result1 -= tmp[i]

    return result1

def two_d_array_reduction_prod_sequential(n):
    shp = (10, 10)
    result1 = np.zeros(shp, np.int_)
    tmp_shp = (n, 10, 10)
    tmp = np.ones(tmp_shp, np.int_)

    for i in range(n):
        tmp[i] = i

    for i in range(n):
        result1 -= tmp[i]

    return result1

print(two_d_array_reduction_prod_sequential(10))
print(njit(two_d_array_reduction_prod_sequential)(10))
print(two_d_array_reduction_prod_parallel(10))
```

```py
$ python test2.py
[[-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]]
[[-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]
 [-45 -45 -45 -45 -45 -45 -45 -45 -45 -45]]
[[45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]
 [45 45 45 45 45 45 45 45 45 45]]
```

I think what might be happening is the reduction, which has intermediate arrays like ([-1, -1, -1...], [-2, -2, -2, ...], ...) is subtracting them instead of adding them, resulting in a double negation. When I set `NUMBA_DEBUG_ARRAY_OPT_RUNTIME=1` the `one_res_print` intermediate reduction arrays are negative integers. (note, you'll want to replace the 10 in the function calls with something less than your number of cores to easily debug this). So either the intermediate arrays should be positive, or the reduction should add instead of subtracting. 

CC @DrTodd13

### PR DESCRIPTION
**Pull request by DrTodd13**
Resolves #4555.

### CODE REVIEW

**Comment by sklam** [comment]
This will fail for:

```python

In [12]: def foo(n):
    ...:     c = n
    ...:     for i in prange(n):
    ...:         c /= 2
    ...:     return c
    ...:

In [13]: njit(parallel=True)(foo)(10)
Out[13]: 976562.5

In [14]: njit(parallel=False)(foo)(10)
Out[14]: 0.009765625
```

File: numba/tests/test_parfors.py [hunk 1]
```diff
@@ -1700,6 +1700,36 @@ def test_impl():
...
+            for i in range(10):
+                s -= 2
+            return s
+        self.prange_tester(test_impl, scheduler_type='unsigned',
+                           check_fastmath=True)
+
+    @skip_unsupported
+    def test_prange03div(self):
+        def test_impl():
+            s = 1.0
```
Comment (sklam) [inline numba/tests/test_parfors.py]: Need to test for case when `s > 1`

Comment (DrTodd13) [reply numba/tests/test_parfors.py]: Improved all the cases a bit.

**Comment by stuartarchibald** [comment]
Thanks for the patch @DrTodd13. @sklam did you have anything further to add on this?

File: numba/npyufunc/parfor.py [hunk 2]
```diff
@@ -366,16 +367,26 @@ def _lower_parfor_parallel(lowerer, parfor):
...
                         if isinstance(rhs, ir.Expr) and rhs.op == 'inplace_binop' and rhs.rhs.name == init_var.name:
+                            if config.DEBUG_ARRAY_OPT:
+                                print("Adding call to reduction", rhs)
+                            if rhs.fn == operator.isub:
+                                rhs.fn = operator.iadd
+                                rhs.immutable_fn = operator.add
+                            if rhs.fn == operator.itruediv or rhs.fn == operator.ifloordiv:
+                                rhs.fn = operator.imul
+                                rhs.immutable_fn = operator.mul
+                            if config.DEBUG_ARRAY_OPT:
```
Comment (stuartarchibald) [inline numba/npyufunc/parfor.py]: Do the corresponding docs need an update: https://github.com/numba/numba/blob/5bd018fd55ab2ff0916ab2c6360334fe94e471a4/docs/source/user/parallel.rst (under "Supported Operations") ?

Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: += and *= are not mentioned in Supported Operations.  They are mentioned in the Explicit Parallel Loops section.  I have updated that section to include -= and /=.

Comment (stuartarchibald) [reply numba/npyufunc/parfor.py]: Great, thanks.

**Comment by stuartarchibald** [comment]
Thanks for the fixes @DrTodd13, looks good to me.

**Pull request merged**